speed up slow quadratic search in addDeferred

MartinNowak opened the pull request:
Roughly 3.5% faster semantic when compiling phobos.

MartinNowak commented on src/dmodule.d:
```diff
@@ -261,11 +261,27 @@ public:
-    extern (C++) static __gshared Dsymbols deferred;    // deferred Dsymbol's needing semantic() run on them
-    extern (C++) static __gshared Dsymbols deferred3;   // deferred Dsymbol's needing semantic3() run on them
+    extern (C++) static __gshared SetDummy[SymHashWrapper] deferred;    // deferred Dsymbol's needing semantic() run on them
+    extern (C++) static __gshared SetDummy[SymHashWrapper] deferred3;   // deferred Dsymbol's needing semantic3() run on them
```
We could alternatively move https://github.com/D-Programming-Language/druntime/pull/1282 into the compiler, to avoid using runtime typeinfo.

MartinNowak commented:
Weird Windows issue, `Error: Internal Compiler Error: unsupported type bool[SymHashWrapper]`.
If we want to preserve order independent of addresses, we could keep the array and add a bloom filter instead. That would also work for the fixup list.

yebblies commented:
I'd guess it's the c++ mangler and the linux one is just fudging it.

MartinNowak commented:
Updated, now uses a bloom filter to preserve the deterministic order of deferred semantic.

9rnsr commented:
The PR subject should be updated - hash set is not used anymore.

The faster deerred semantic analysis would also be usefful for #5500.

MartinNowak commented:
Well the interesting thing is that there aren't any duplicate symbols being added when compiling phobos, so the question is under which circumstances duplicates might arise, and if we could turn this loop into a debug assertion instead.

yebblies commented:
If duplicates are rare, would we be better off just appending and skipping symbols that already have semantic run when we get to them?

MartinNowak commented:
> If duplicates are rare, would we be better off just appending and skipping symbols that already have semantic run when we get to them?

Will try that.

MartinNowak committed:
speed up slow quadratic search in addDeferred

- don't know how to generate duplicates, but they should be really rare
  and even then symbols cache their semanticPass state, so duplicates
  wouldn't even be analyzed twice

MartinNowak commented:
> Will try that.

Should work @yebblies.

yebblies commented:
Nice!

yebblies commented:
Auto-merge toggled on
